event/monitoring.go

```diff
@@ -89,3 +92,75 @@ type PoolEvent struct {
 type PoolMonitor struct {
 	Event func(*PoolEvent)
 }
+
+// ServerDescriptionChangedEvent represents a server description change.
+type ServerDescriptionChangedEvent struct {
+	Address             address.Address
+	ID                  primitive.ObjectID
+	PreviousDescription description.Server
+	NewDescription      description.Server
+}
+
+// ServerOpeningEvent is an event generated when the server is initialized.
+type ServerOpeningEvent struct {
+	Address address.Address
+	ID      primitive.ObjectID
+}
+
+// ServerClosedEvent is an event generated when the server is closed.
+type ServerClosedEvent struct {
+	Address address.Address
+	ID      primitive.ObjectID
+}
+
+// TopologyDescriptionChangedEvent represents a topology description change.
+type TopologyDescriptionChangedEvent struct {
+	ID                  primitive.ObjectID
+	PreviousDescription description.Topology
+	NewDescription      description.Topology
+}
+
+// TopologyOpeningEvent is an event generated when the topology is initialized.
+type TopologyOpeningEvent struct {
+	ID primitive.ObjectID
+}
+
+// TopologyClosedEvent is an event generated when the topology is closed.
+type TopologyClosedEvent struct {
+	ID primitive.ObjectID
+}
+
+// ServerHeartbeatStartedEvent is an event generated when the ismaster command is started.
+type ServerHeartbeatStartedEvent struct {
+	ConnectionID string
+	Awaited      bool
+}
+
+// ServerHeartbeatSucceededEvent is an event generated when the ismaster succeeds.
+type ServerHeartbeatSucceededEvent struct {
+	Duration     int64
+	Reply        description.Server
+	ConnectionID string
+	Awaited      bool
+}
+
+// ServerHeartbeatFailedEvent is an event generated when the ismaster fails.
+type ServerHeartbeatFailedEvent struct {
+	Duration     int64
+	Reply        error
+	ConnectionID string
+	Awaited      bool
+}
+
+// SdamMonitor represents a monitor that is triggered for different SDAM events.
```

Comment:
This type of monitoring is pretty complex and requires understanding all of the stuff that goes on in the driver's background routines to write effective handlers for the events. Can you add some more high-level documentation here or in `ClientOptions` and file a follow-on ticket to add a `event/doc.go` file that explains all of the different types of monitoring in more detail?

Reply:
this could probably be improved. created GODRIVER-1747